go/merge/three_way.go

```diff
@@ -200,7 +190,81 @@ func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter, p
 		}
 		panic(fmt.Errorf("Bad key type in diff: %s", change.V.Type().Describe()))
 	}
-	return threeWayOrderedSequenceMerge(parent, aDiff, bDiff, makeGetFunc(a), makeGetFunc(b), makeGetFunc(parent), apply, vwr, progress)
+	return m.threeWayOrderedSequenceMerge(structCandidate{a}, structCandidate{b}, structCandidate{parent}, apply, path)
+}
+
+type candidate interface {
```

Comment:
I see. Yes that would get pretty noisy, though I also find the adaptors quite distracting. What about another entirely different approach: if merger had a map of Kind --> getter/diff functions instead? (in fact it could be global)